core/moddeps.go

```diff
@@ -0,0 +1,215 @@
+package core
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/dagger/dagger/cmd/codegen/introspection"
+	"github.com/dagger/dagger/dagql"
+	dagintro "github.com/dagger/dagger/dagql/introspection"
+	"github.com/dagger/dagger/tracing"
+)
+
+const (
+	modMetaDirPath    = "/.daggermod"
+	modMetaOutputPath = "output.json"
+
+	ModuleName = "daggercore"
+)
+
+/*
+ModDeps represents a set of dependencies for a module or for a caller depending on a
+particular set of modules to be served.
+*/
+type ModDeps struct {
+	root *Query
+	Mods []Mod // TODO hide
+
+	// should not be read directly, call Schema and SchemaIntrospectionJSON instead
+	lazilyLoadedSchema            *dagql.Server
+	lazilyLoadedIntrospectionJSON string
+	loadSchemaErr                 error
+	loadSchemaLock                sync.Mutex
+}
+
+func NewModDeps(root *Query, mods []Mod) *ModDeps {
+	return &ModDeps{
+		root: root,
+		Mods: mods,
+	}
+}
+
+func (d *ModDeps) Prepend(mods ...Mod) *ModDeps {
+	deps := append([]Mod{}, mods...)
+	deps = append(deps, d.Mods...)
+	return NewModDeps(d.root, deps)
+}
+
+func (d *ModDeps) Append(mods ...Mod) *ModDeps {
+	deps := append([]Mod{}, d.Mods...)
+	deps = append(deps, mods...)
+	return NewModDeps(d.root, deps)
+}
```

Comment:
It might not actually matter - but I noticed we were sometimes appending and sometimes prepending and assumed it might be significant, so I just did a quick refactor so I could identify it more easily while figuring out how everything worked.

This kinda sorta feels like a delegation hierarchy to me, and usually those are in some kind of precedence order, but yeah given how everything is namespaced without any overlap it really shouldn't matter.